smolyay/basis.py

```diff
@@ -403,29 +482,393 @@ def _derivative(self, x):
+        Basis functions must have the same domain.
     """
 
     def __init__(self, basis_functions):
+        if len(basis_functions) == 0:
+            raise IndexError("Must have at least one basis function.")
+        domain = basis_functions[0].domain
+        if any(not numpy.array_equal(domain, b.domain) for b in basis_functions[1:]):
+            raise ValueError("Basis functions must have the same domain.")
         self._basis_functions = basis_functions
 
+
+class MutableBasisFunctionSet(BasisFunctionSet):
+    """Set of basis functions where number of functions is mutable.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def __init__(self, num_terms):
+        self._num_terms = None
+        self._basis_functions = None
+
+        self.num_terms = num_terms
+
     @property
     def basis_functions(self):
         """list: Basis functions."""
+        if self._basis_functions is None:
+            self._create()
         return self._basis_functions
+
+    @property
+    def num_terms(self):
+        """int: the number of terms in the set."""
+        return self._num_terms
+
+    @num_terms.setter
+    def num_terms(self, value):
+        num_terms = int(value)
+        if num_terms <= 0:
+            raise ValueError("Must have at least one term.")
+        if num_terms != self._num_terms:
+            self._num_terms = num_terms
+            self._basis_functions = None
+
+    @abc.abstractmethod
+    def _create(self):
+        """Create the basis functions in the set."""
+        pass
+
+
+class ChebyshevFirstKindBasisFunctionSet(MutableBasisFunctionSet):
+    """Set of Chebyshev polynomials of the first kind.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def _create(self):
+        """Create the basis functions in the set."""
+        self._basis_functions = [ChebyshevFirstKind(f) for f in range(self._num_terms)]
+
+
+class ChebyshevSecondKindBasisFunctionSet(MutableBasisFunctionSet):
+    """Set of Chebyshev polynomials of the second kind.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def _create(self):
+        """Create the basis functions in the set."""
+        self._basis_functions = [ChebyshevSecondKind(f) for f in range(self._num_terms)]
+
+
+class TrigonometricBasisFunctionSet(MutableBasisFunctionSet):
+    """Set of Trigonmetric equations.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def _create(self):
+        """Create the basis functions in the set."""
+        index_trig = numpy.arange(self.num_terms, dtype=int)
+        frequencies = numpy.where(
+            index_trig % 2 == 1, (1 + index_trig) // 2, -index_trig // 2
+        )
+        self._basis_functions = [Trigonometric(f) for f in frequencies]
+
+
+class NestedBasisFunctionSet(BasisFunctionSet):
```

Comment:
Have this class's `__init__` take its list of basis functions, plus `num_per_level`. Calculate `start_level` and `end_level` for that.